src/main/java/org/folio/inventory/dataimport/handlers/actions/CreateMarcHoldingsEventHandler.java

```diff
@@ -0,0 +1,228 @@
+      return future;
+      });
+    } catch (Exception e) {
+      LOGGER.error("Failed to create Holdings", e);
+      future.completeExceptionally(e);
+    }
+    return future;
+  }
+
+  private void defaultMapRecordToHoldings(DataImportEventPayload dataImportEventPayload) {
+    try {
+      HashMap<String, String> context = dataImportEventPayload.getContext();
+      var mappingRules = new JsonObject(context.get(MAPPING_RULES_KEY));
+      var parsedRecord = new JsonObject((String) new JsonObject(context.get(MARC_HOLDINGS.value()))
+        .mapTo(Record.class).getParsedRecord().getContent());
+      var mappingParameters = new JsonObject(context.get(MAPPING_PARAMS_KEY)).mapTo(MappingParameters.class);
+      RecordMapper<Holdings> recordMapper = RecordMapperBuilder.buildMapper(MARC_FORMAT);
+      var holdings = recordMapper.mapRecord(parsedRecord, mappingParameters, mappingRules);
+      dataImportEventPayload.getContext().put(HOLDINGS.value(), Json.encode(new JsonObject().put(HOLDINGS_PATH, JsonObject.mapFrom(holdings))));
+    } catch (Exception e) {
+      LOGGER.error("Failed to map Record to Holdings", e);
+      throw new JsonMappingException("Error in default mapper.", e);
+    }
+  }
+
+  private CompletableFuture<String> findInstanceIdByHrid(DataImportEventPayload dataImportEventPayload, JsonObject holdingAsJson) throws JsonProcessingException {
+    CompletableFuture<String> future = new CompletableFuture<>();
+    if (StringUtils.isBlank(holdingAsJson.getString(INSTANCE_ID_FIELD))) {
+      var recordAsString = dataImportEventPayload.getContext().get("MARC_HOLDINGS");
+      var record = ObjectMapperTool.getMapper().readValue(recordAsString, Record.class);
+      var instanceHrid = getControlFieldValue(record, "004");
+
+      if (isBlank(instanceHrid)) {
+        throw new EventProcessingException(FIELD_004_MARC_HOLDINGS_NOT_NULL);
+      }
+
+      var context = EventHandlingUtil.constructContext(dataImportEventPayload.getTenant(), dataImportEventPayload.getToken(), dataImportEventPayload.getOkapiUrl());
+      var instanceCollection = storage.getInstanceCollection(context);
+
+      try {
+        instanceCollection.findByCql(format("hrid=%s", instanceHrid), PagingParameters.defaults(),
+          findResult -> {
+            String instanceId = null;
+            if (findResult.getResult() != null && findResult.getResult().totalRecords == 1) {
+              var records = findResult.getResult().records;
+              var instance = records.stream()
+                .findFirst()
+                .orElseThrow(() -> new EventProcessingException("No instance id found for marc holdings with hrid: " + instanceHrid));
+              instanceId = instance.getId();
+            }
+            future.complete(instanceId);
+          },
+          failure -> {
+            LOGGER.error(ERROR_HOLDING_MSG + format(". StatusCode: %s. Message: %s", failure.getStatusCode(), failure.getReason()));
+            future.complete(EMPTY);
```

Comment:
Maybe will be better to complete the future exceptionally in case of this failure?